libraries/Netdump/src/NetdumpPacket.cpp

```diff
@@ -0,0 +1,251 @@
+    This file is part of the esp8266 core for Arduino environment.
+
+    This library is free software; you can redistribute it and/or
+    modify it under the terms of the GNU Lesser General Public
+    License as published by the Free Software Foundation; either
+    version 2.1 of the License, or (at your option) any later version.
+
+    This library is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+    Lesser General Public License for more details.
+
+    You should have received a copy of the GNU Lesser General Public
+    License along with this library; if not, write to the Free Software
+    Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+*/
+
+#include "Netdump.h"
+#include <lwip/init.h>
+
+constexpr char* NetdumpPacket::packetTypeArray[];
+
+void NetdumpPacket::printDetail(Print& out, const String& indent, const char* data, size_t size, PacketDetail pd) const
+{
+    if (pd == PacketDetail::NONE)
+    {
+        return;
+    }
+
+    uint16_t charCount = pd == PacketDetail::FULL ? 24 : 80;
+
+    size_t start = 0;
+    while (start < size)
+    {
+        size_t end = start + charCount;
+        if (end > size)
+        {
+            end = size;
+        }
+        out.printf("%s", indent.c_str());
+        if (pd == PacketDetail::FULL)
+        {
+            for (size_t i = start; i < end; i++)
+            {
+                out.printf("%02x ", (unsigned char)data[i]);
+            }
+            for (size_t i = end; i < start + charCount; i++)
+            {
+                out.print("   ");
+            }
+        }
+        for (size_t i = start; i < end; i++)
+        {
+            out.printf("%c", data[i] >= 32 && data[i] < 128 ? data[i] : '.');
+        }
+        out.println();
+
+        start += charCount;
+    }
+}
+
+NetdumpPacket::PacketType NetdumpPacket::packetType() const
+{
+	if (isARP()) return PacketType::ARP;
+	if (isIP())
+	{
+		if (isUDP())
+		{
+			if (isMDNS()) return PacketType::MDNS;
+			if (isDNS())  return PacketType::DNS;
+			if (isSSDP()) return PacketType::SSDP;
+			if (isDHCP()) return PacketType::DHCP;
+			if (isWSDD()) return PacketType::WSDD;
+			if (isNETBIOS()) return PacketType::NETBIOS;
+			if (isSMB())  return PacketType::SMB;
+			if (isOTA())  return PacketType::OTA;
+			return PacketType::UDP;
+		}
+		if (isTCP())
+		{
+			if (isHTTP()) return PacketType::HTTP;
+			return PacketType::TCP;
+		}
+		if (isICMP()) return PacketType::ICMP;
+		if (isIGMP()) return PacketType::IGMP;
+		return PacketType::IP;
+	}
+	return PacketType::UKNW;
+}
+
+
+String NetdumpPacket::toString(PacketDetail netdumpDetail) const
+{
+    StreamString sstr;
+    sstr.reserve(128);
+
+    sstr.printf("%d %3s %s ", netif_idx, out ? "out" : "in ", packetTypeString(packetType()));
+
+    switch (packetType())
```

Comment:
Nested switch statements are bad practice for several reasons. It is better to encapsulate with functions, e.g.:
```cpp
switch(packetType)
{
  case A:
  {
    sstr = packetCaseA();
    return sstr;
  }
  case B:
  {
    sstr = packetCaseB();
    return sstr;
  }
  case C:
  case D:
  {
    sstr = packetCasesCD();
    return sstr;
  }
  default:
  {
    sstr = packetDefault();
    return sstr
  }
}
...

StreamString packetCaseA()
{
  switch(getARPType())
  {
  case 1:
    sstr = arpCase1();
    return sstr;
  }
...
}
```

Also:
- make sure there is always a default handling case in every switch
- no code on the case line (code lines should go below the case)